File: leetcode-solutions/ruby/203. Remove Linked List Elements.rb
# https://leetcode.com/problems/remove-linked-list-elements/

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} val
# @return {ListNode}
def remove_elements(head, val)
  prev = nil
  curr = head
  
  while curr != nil
      if curr.val == val
          # need to remove head
          if prev == nil
              curr = curr.next
              head = curr
          else
              prev.next = curr.next
              curr = curr.next
          end
      else
          prev = curr
          curr = curr.next
      end
  end
  
  head
end

File: leetcode-solutions/python/203. Remove Linked List Elements.py
# https://leetcode.com/problems/remove-linked-list-elements/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        new_head = ListNode()
        new_tail = new_head

        while head != None:
            if head.val != val:
                new_tail.next = head
                new_tail = new_tail.next
            head = head.next
        new_tail.next = None
        return new_head.next
